---
title: "11__KSEA__3"
output: html_document
---


Background

Here I do KSEA to find insulin-regulated and defective kinases

I also explore MARK2 and MARK3







###Initialise
```{r}
library(ksea)
library(reshape2)
library(ggplot2)
library(rlist)
source("..\\scripts/essential_visualisations_JvG.R")
source("..\\scripts/enrichment_analysis_JvG.R")
source("..\\scripts/3T3L1_mouse_IR__DF__202104_objects.R")
source("..\\scripts/3T3L1_mouse_IR__DF__202104_functions.R")
```

###Load in data
```{r}
load("data/intermediate/phos_3t3/phos_3t3_proc__4.RData")
load("data/intermediate/prot_3t3/proteome_post_analysis_3t3_DIA_matched__1.RData")


###Kinase substrate
psp_kinase_substrate <- read.delim("data/intermediate/psp/Kinase_Substrate_Dataset")
psp_kinase_substrate$uniprot_site <- apply(psp_kinase_substrate[, c("SUB_ACC_ID", 
                                    "SUB_MOD_RSD")],
      1,
      function(row){
        return(paste(row,
              collapse = "_"))
      })
```



####Fusing kinase isoforms
```{r}
#Here I pool kinase isoforoms
kin_merger <- function(kinases_pooled){
  kinases_pooled[grep("^Akt", kinases_pooled)] <- "Akt"

  kinases_pooled[grep("^AMPK", kinases_pooled)] <- "AMPK"

  kinases_pooled[grep("^Aur", kinases_pooled)] <- "Aur"

  kinases_pooled[grep("^CAMK2", kinases_pooled)] <- "CAMK2"

  kinases_pooled[grep("^Chk", kinases_pooled)] <- "Chk"

  kinases_pooled[grep("^CK1", kinases_pooled)] <- "CK1"

  kinases_pooled[grep("^CK2", kinases_pooled)] <- "CK2"

  kinases_pooled[grep("^ERK", kinases_pooled)] <- "ERK"

  kinases_pooled[grep("^GSK3", kinases_pooled)] <- "GSK3"

  kinases_pooled[grep("^HIPK", kinases_pooled)] <- "HIPK"

  kinases_pooled[grep("^IKK", kinases_pooled)] <- "IKK"
  
  kinases_pooled[grep("^JNK", kinases_pooled)] <- "JNK"

  kinases_pooled[grep("^KHS", kinases_pooled)] <- "KHS"

  kinases_pooled[grep("^LATS", kinases_pooled)] <- "LATS"

  kinases_pooled[grep("^LIMK", kinases_pooled)] <- "LIMK"

  kinases_pooled[grep("^MARK", kinases_pooled)] <- "MARK"

  kinases_pooled[grep("^MKK", kinases_pooled)] <- "MKK"
  
  kinases_pooled[grep("^MSK", kinases_pooled)] <- "MSK"

  kinases_pooled[grep("^NDR", kinases_pooled)] <- "NDR"

  kinases_pooled[grep("^NEK", kinases_pooled)] <- "NEK"

  kinases_pooled[grep("^P38", kinases_pooled)] <- "P38"

  kinases_pooled[grep("^p70S6K", kinases_pooled)] <- "p70S6K"

  kinases_pooled[grep("^P70S6K", kinases_pooled)] <- "p70S6K"

  kinases_pooled[grep("^PAK", kinases_pooled)] <- "PAK"

  kinases_pooled[grep("^PDHK", kinases_pooled)] <- "PDHK"

  kinases_pooled[grep("^Pim", kinases_pooled)] <- "Pim"

  kinases_pooled[grep("^PKC", kinases_pooled)] <- "PKC"

  kinases_pooled[grep("^PKG", kinases_pooled)] <- "PKG"

  kinases_pooled[grep("^PRKD", kinases_pooled)] <- "PRKD"

  kinases_pooled[grep("^ROCK", kinases_pooled)] <- "ROCK"

  kinases_pooled[grep("^SGK", kinases_pooled)] <- "SGK"

  kinases_pooled[grep("^ULK", kinases_pooled)] <- "ULK"
  

  return(kinases_pooled)
}

psp_kinase_substrate$KINASE_merged <-  kin_merger(psp_kinase_substrate$KINASE)
psp_kinase_substrate$KINASE_merged <- toupper(psp_kinase_substrate$KINASE_merged)

###Inspect kinases that have >= 10 substrates in our data
#They all apeear to be fused where necessary
merged_kinases <- unique(psp_kinase_substrate$KINASE_merged)
merged_kinases_numsubs <- sapply(merged_kinases, 
                                 function(kinase){
                                   subs <- unique(psp_kinase_substrate$uniprot_site[which(psp_kinase_substrate$KINASE_merged == 
                                                                                            kinase)])
                                   return(length(which(phos_3t3_proc$uniprot_site %in% subs)))
                                 })
sort(merged_kinases_numsubs, decreasing = TRUE)

##Make new df
psp_kinase_substrate_merged <- psp_kinase_substrate
psp_kinase_substrate_merged$KINASE <- psp_kinase_substrate_merged$KINASE_merged
```


###KSEA

INS/BAS
```{r, fig.show = "hide"}
#Only kinaes with 10 or more subs
set.seed(1413)
ksea_beltrao_kinmerged_output <- ksea_beltrao_enricher_mult_conditions(phos_3t3_proc,
                                                                       psp_kinase_substrate_merged,
                                                             c("CTRL",
                                                                  "CI",
                                                                  "DEX",
                                                                  "TNF",
                                                                  "MPQ",
                                                                  "AA"),
                                                             min_sub_num = 10)


###Output
save(ksea_beltrao_kinmerged_output,
     file = "output/data/kinases/KSEA/ksea_beltrao_kinmerged_output.RData")
```


#####Visualise
```{r}
###Function: Stacked and signed barplot for KSEA with multiple conditions
#Only incorporates ES data, not significance
KSEA_stacked_barplotter <- function(data,
                                    order_col,
                                    ES_cols,
                                    pval_cols,
                                    colours,
                                    #sig_only does nothing
                                    sig_only = FALSE,
                                    sig_cutoff = 0.05,
                                    return_df = FALSE){


data$kinase <- rownames(data)

##Order
data <- data[order(data[, order_col]), ]


##Melt and process
#ES
data_melt_ES <- melt(data,
                  measure.vars = ES_cols,
                  id.vars = "kinase")
data_melt_ES$ES <- data_melt_ES$value
#pval
data_melt_pval <- melt(data,
                       measure.vars = pval_cols,
                       id.vars = "kinase")
data_melt_pval$pval <- data_melt_pval$value
#combine
data_melt <- data_melt_ES
data_melt$pval <- data_melt_pval$pval




#condition as factor
data_melt$condition <- factor(gsub("_ES",
                                   "",
                                   data_melt$variable),
                              levels = gsub("_ES",
                                            "",
                                            ES_cols))

#kinase as factor
data_melt$kinase <- factor(data_melt$kinase,
                           levels = unique(data_melt$kinase))


#Make sig only df
data_melt_sig_only <- data_melt
data_melt_sig_only$ES[which(data_melt_sig_only$pval >= sig_cutoff)] <- NA


##Plot
output_plot <- ggplot() +
  geom_col(data = data_melt,
           aes(x = kinase,
               y = ES,
               colour = condition),
           position  = position_dodge(width = 0.65,
                                      preserve = "single"),
           orientation = "x",
           width = 0.45,
           fill = "white",
           size = 0.275) +
   geom_col(data = data_melt_sig_only,
           aes(x = kinase,
               y = ES,
               fill = condition),
           position  = position_dodge(width = 0.65,
                                      preserve = "single"),
           orientation = "x",
           width = 0.45) +
  geom_hline(yintercept = 0) +
  scale_colour_manual(values = colours,
                    name = "Model") + 
  scale_fill_manual(values = colours,
                    name = "Model") +
  theme(panel.background = element_blank(),
        panel.grid = element_blank(),
        axis.line.y = element_line(colour = "black"),
        axis.text.x = element_text(angle = 90,
                                   vjust = 0.5,
                                   hjust = 1,
                                   colour = "black"),
        axis.text.y = element_text(colour = "black"),
        plot.title = element_text(hjust = 0.5)) +
  labs(x = "Kinase",
       y = "Enrichment score")

if (return_df){
  
  return(data_melt)
} else {
  return(output_plot)
}
}


#Function: Get number of subsrates for each kinase in a psp_kinase_substrate database
#psp_kinase_substrate: Phosphositeplus kinase substrate database including a uniprot_site column to identify substrates
#phos_data: Phospho dataset including a uniprot_site column to identify substrates
psp_kinsubnums <- function(psp_kinase_substrate,
                              phos_data){
  #Get kinases
  kinases <- unique(psp_kinase_substrate$KINASE)
  #Get kinase sub nums
  kinase_subnums <- sapply(kinases,
                           function(kinase){
                             kinase_subs <- psp_kinase_substrate$uniprot_site[which(psp_kinase_substrate$KINASE == kinase)]
                             return(length(which(phos_data$uniprot_site %in% kinase_subs)))
                           })
  return(kinase_subnums)
}

#Function: Get kinases that have a minimum number of substrates in your database
#psp_kinase_substrate: Phosphositeplus kinase substrate database including a uniprot_site column to identify substrates
#phos_data: Phospho dataset including a uniprot_site column to identify substrates
#min_num_subs: Minimum nubmer of subsrates required
psp_kinsub_filter <- function(psp_kinase_substrate,
                              phos_data,
                              min_subs = 10){
  #Get number of substrates per kinase
  kinase_subnums <- psp_kinsubnums(psp_kinase_substrate = psp_kinase_substrate,
                                   phos_data = phos_data)
  return(names(kinase_subnums)[which(kinase_subnums >= min_subs)])
}
```


#####Stacked barplots

INS/BAS
```{r}
##Only significant kinases
KSEA_stacked_barplotter(data  = ksea_beltrao_kinmerged_output[which(rowSums(ksea_beltrao_kinmerged_output[, grep("adj_pval",
                                     colnames(ksea_beltrao_kinmerged_output))] < 0.05) > 0), ],
                        order_col = "CTRL_ES",
                        ES_cols = c("CTRL_ES",
                                   "CI_ES",
                                   "DEX_ES",
                                   "TNF_ES",
                                   "MPQ_ES",
                                   "AA_ES"),
                        pval_cols = c("CTRL_adj_pval",
                                   "CI_adj_pval",
                                   "DEX_adj_pval",
                                   "TNF_adj_pval",
                                   "MPQ_adj_pval",
                                   "AA_adj_pval"),
                        colours = SH_control_model_main_colours) +
  ggtitle("Significant Kinases") 
  ggsave("output/images/kinases/ksea/ksea_beltrao_kinmerged_output_sig.pdf",
         height = 4,
         width = 6)
```




####Follow up kinases

###MARK2/3
```{r}
##Initialise
library(tidyverse)
library(purrr)
source("..\\scripts/PSP_JvG.R")
#Load in PSP data
psp_phos_data <- load_psp_data("PSP__phos_dataset__20210127",
                               directory = "data/intermediate/psp/")

###Get all psp MARK2/3 substrates
psp_kinase_substrate_MARK23 <- psp_kinase_substrate %>%
  subset(toupper(KINASE) %in% c("MARK2", "MARK3"))
psp_kinase_substrate_MARK2 <- psp_kinase_substrate %>%
  subset(toupper(KINASE) == "MARK2")
psp_kinase_substrate_MARK3 <- psp_kinase_substrate %>%
  subset(toupper(KINASE) == "MARK3")

#Get all relevant 3t3l1 psites
phos_3t3_proc_MARK23relevant <- phos_3t3_proc %>%
  subset(tolower(gene) %in% tolower(psp_kinase_substrate_MARK23$SUB_GENE))
#Get SITE_GRP_IDs for these
phos_3t3_proc_MARK23relevant$SITE_GRP_ID <- 
  unlist(uniprotsite_to_sitegroupid(phos_3t3_proc_MARK23relevant$uniprot_site,
                                    psp_phos_data = psp_phos_data))

#Select those that are substrates
#Build a list of substrates of MARK2 and 3, mark2, and mark3
phos_3t3_proc_MARK23_list <- map(list("MARK23" = psp_kinase_substrate_MARK23,
                                 "MARK2" = psp_kinase_substrate_MARK2,
                                 "MARK3" = psp_kinase_substrate_MARK3),
                            ~ phos_3t3_proc_MARK23relevant[which(phos_3t3_proc_MARK23relevant$SITE_GRP_ID %in%
                                                                   .$SITE_GRP_ID), ])


##How do subs correlate with our MARK2/3 defective psites?
#Get MARK2/3 defective psites
MARK23_def_psites <- phos_3t3_proc %>%
  subset(num_defective_models >= 4) %>%
  subset(gene == "Mark2" | gene == "Mark3") %>%
  rownames()
names(MARK23_def_psites) <- toupper(gsub("_.+", "", MARK23_def_psites))

#Function for correlation
correlate_target_bait <- function(target_vec,
                                  bait_data){
  target_vec <- as.numeric(target_vec)
  bait_data <- as.matrix(bait_data)
  corr_output <- apply(bait_data,
      1,
      function(x){
        temp_test <- cor.test(target_vec, x)
        return(c("r" = temp_test$estimate,
                 "p" = temp_test$p.value))
      })
  return(corr_output)
}


protnorm_cols <- phos_3t3_proc %>%
  .[, grep("PROT_norm_\\d+", colnames(.))] %>% colnames()
protnorm_cols_unnorm <- gsub("PROT_norm_", "", protnorm_cols)

#Run correlations
corr_args <- cross(list(protein = list("MARK2", "MARK3"),
      protnorm = list("protnorm", "unnorm")))
MARK23_def_psites_subcors <- map(corr_args, function(x){
      #Protnorm if desired
      if (x$protnorm == "protnorm"){
        bait_data <- phos_3t3_proc_MARK23_list[[1]][, protnorm_cols]
        target_vec <- phos_3t3_proc[MARK23_def_psites[x$protein], protnorm_cols_unnorm]
      } else {
        bait_data <- phos_3t3_proc_MARK23_list[[1]][, 1:52]
        target_vec <- phos_3t3_proc[MARK23_def_psites[x$protein], 1:52]
      }
      #Run corrs
      output_data <- correlate_target_bait(target_vec = target_vec, bait_data = bait_data)
      rownames(output_data) <- c("r", "p")
      return(output_data)
    })
names(MARK23_def_psites_subcors) <- map(corr_args, ~ paste(unlist(.), collapse = "_"))
#Collapse into a single df
MARK23_def_psites_subcors_df <- map2(MARK23_def_psites_subcors, names(MARK23_def_psites_subcors),
     function(data, name){
       data <- t(data)
       colnames(data) <- paste(name, colnames(data), sep = "_")
       return(data)
     }) %>%
  purrr::reduce(cbind) %>%
  as.data.frame()

###Heatmap of correlations
#unnorm  data only
fc_sig_heatmap(data = MARK23_def_psites_subcors_df,
                 fc_cols_w_pvals = c("MARK2_unnorm_r",
                                     "MARK3_unnorm_r"),
                 pval_cols = c("MARK2_unnorm_p",
                                     "MARK3_unnorm_p"),
                 order_column = "MARK2_unnorm_r",
                 is_decreasing = FALSE,
                 x_axis_names = c("MARK2",
                                  "MARK3"),
               legend_title = "R")
ggsave_pdfpng("output/images/kinases/MARK_2_3/hmap_MARK23_defsites_subs_cor",
              width = 2, height = 2)

#Export correlations
write.csv(MARK23_def_psites_subcors_df,
          file = "output/data/kinases/MARK23/MARK23_def_psites_subcors_df.csv")
```





























